### PR TITLE
Issue #86: Reduce axclustermanager container size.

### PR DESCRIPTION
**Pull request by lcjlcj**
This reduces axclustermanager (argocluster) container from 1.5GB to less than 600MB. There are two main changes:

- Some additional cleanup like apt-get packages, pyc files, old kubectl, docker daemons, pip install caches, google sdk install. This takes off about 500MB. And it was relatively straightforward.
- Kubernetes server binary was embedded with this container when we had kubernetes customizations. It's not needed any more now that we download it directly from kubernetes release.

We still have customized kubernetes salt. kube-up is modified to take customized salt from container but download server from kubernetes. SERVER_BINARY_TAR_URL and SERVER_BINARY_TAR_HASH are not set by kube up but passed in from caller of kube-up.

Kubernetes distro version was complex to reflect customizations. This is not needed now. Simplify naming scheme to match actual kubernetes release.

Hacks at upgrade time to fix up AWS user data should be re-written anyway. But for now just fix them to handle new version format. Upgrade also search for customized version signatures and is not needed and removed.

### CODE REVIEW

File: platform/source/lib/ax/platform/ax_master_manager.py [hunk 1]
```diff
@@ -162,7 +164,7 @@ def get_user_data(self):
         try:
             user_data = self.user_data_fixup(user_data)
         except Exception as e:
-            raise AXPlatformException("Failed while fixing up user-data: " + str(e))
+            raise AXPlatformException("Failed while fixing up user-data: " + traceback.format_exc(e))
```
Comment (zhan849) [inline platform/source/lib/ax/platform/ax_master_manager.py]: logger.exception?

File: platform/containers/kubernetes-distro/Dockerfile [hunk 2]
```diff
@@ -5,8 +5,9 @@ ENV DOCKER_VERSION 1.12.6
 ENV DOCKER_SHA256 cadc6025c841e034506703a06cf54204e51d0cadfae4bae62628ac648d82efdd
 # Current version.
 ENV KUBERNETES_VERSION=1.6.7
+ENV KUBERNETES_SERVER_HASH=29828b8913fc0d922f42cb7b04d15a518984e0d6
 # All supported versions, including backward compatible versions.
-ENV KUBERNETES_VERSIONS_LIST="1.5.7 1.6.7"
+ENV KUBERNETES_VERSIONS_LIST="1.6.7"
```
Comment (zhan849) [inline platform/containers/kubernetes-distro/Dockerfile]: Can we get server binary URL as input, which will leave space for customized kubernetes build.

(If this'd been the case, software_info package related things needs to be changed as well)

File: platform/containers/kubernetes-distro/install.sh [hunk 3]
```diff
@@ -16,25 +16,19 @@ apt-get update && apt-get install -y \
...
 
 pip install -r /tmp/requirements.txt
 pip install awscli
 
 curl -fSL "https://${DOCKER_BUCKET}/builds/Linux/x86_64/docker-$DOCKER_VERSION.tgz" -o docker.tgz \
     && echo "${DOCKER_SHA256} *docker.tgz" | sha256sum -c - \
     && tar -xzvf docker.tgz \
-    && mv docker/* /usr/local/bin/ \
-    && rmdir docker \
+    && mv docker/docker /usr/local/bin/ \
```
Comment (zhan849) [inline platform/containers/kubernetes-distro/install.sh]: we can remove entire docker binary as well as we are not calling docker to upgrade (another 20M reduce?)

File: platform/containers/axclustermanager/Dockerfile [hunk 4]
```diff
@@ -1,10 +1,10 @@
-FROM %%ARGO_BASE_REGISTRY%%/argobase/kubernetes-distro:ax-v1.6.7-6341-095136c-050831f
+FROM %%ARGO_BASE_REGISTRY%%/argobase/kubernetes-distro:v1.6.7
 
 # Remove built-in kubeup code and copy AX version to container.
 RUN rm -rf /kubernetes/cluster
 ADD docker_build/cluster /kubernetes/cluster
 
-ENV AX_KUBE_VERSION ax-v1.6.7-6341-095136c-050831f
+ENV AX_KUBE_VERSION ${KUBERNETES_VERSION}
```
Comment (shrinandj) [inline platform/containers/axclustermanager/Dockerfile]: ${KUBERNETES_VERSION} is an env variable. Will this get substituted when we run docker build?

File: bash-helpers.sh [hunk 5]
```diff
@@ -45,10 +45,10 @@ kdef()
         return 1
     fi
 
-    if [[ -z "$(which kubectl-1.5.7)" || -z "$(which kubectl-1.6.7)" ]]
+    if [[ -z "$(which kubectl-1.6.7)" ]]
```
[the remark this answers is not in the document]
Comment (lcjlcj) [reply bash-helpers.sh]: We should use 1.6.7 or newer version for GCP.

File: platform/containers/kubernetes-distro/Dockerfile (hunk 2 above)
Comment (lcjlcj) [reply platform/containers/kubernetes-distro/Dockerfile]: Yes. That is possible. But I'd like to discourage customized kubernetes build.

File: platform/containers/axclustermanager/Dockerfile (hunk 4 above)
Comment (zhan849) [reply platform/containers/axclustermanager/Dockerfile]: @shrinandj it's set in base image

File: platform/containers/kubernetes-distro/install.sh (hunk 3 above)
Comment (lcjlcj) [reply platform/containers/kubernetes-distro/install.sh]: Docker client is such a handy tool that I'd prefer to keep. We are not worrying about extra 20MB now. There are other ways to get rid of 20MB.

File: platform/source/lib/ax/platform/ax_master_manager.py (hunk 1 above)
Comment (lcjlcj) [reply platform/source/lib/ax/platform/ax_master_manager.py]: Fixed.

File: bash-helpers.sh [hunk 6]
```diff
@@ -71,7 +71,7 @@ kdef()
         export KUBECTL=kubectl-1.5.7
     else
         echo "Target cluster ${DEFAULT_KCLUSTER} in AWS cloud."
-        server=`kubectl-1.5.7 --kubeconfig=${HOME}/.kube/cluster_${DEFAULT_KCLUSTER}.conf version | grep -i server | cut -d ":" -f 5 2> /dev/null`
+        server=`kubectl-1.6.7 --kubeconfig=${HOME}/.kube/cluster_${DEFAULT_KCLUSTER}.conf version | grep -i server | cut -d ":" -f 5 2> /dev/null`
         if [[ $server == *"v1.6"* ]]; then
             export KUBECTL=kubectl-1.6.7
```
[the remark this answers is not in the document]
Comment (lcjlcj) [reply bash-helpers.sh]: Fixed.

File: bash-helpers.sh [hunk 7]
```diff
@@ -71,7 +71,7 @@ kdef()
         export KUBECTL=kubectl-1.5.7
```
[the remark this answers is not in the document]
Comment (lcjlcj) [reply bash-helpers.sh]: Fixed.

File: platform/containers/axclustermanager/Dockerfile (hunk 4 above)
Comment (lcjlcj) [reply platform/containers/axclustermanager/Dockerfile]: @shrinandj good question. I had doubt and tested it. It worked.

File: platform/containers/kubernetes-distro/install.sh (hunk 3 above)
Comment (zhan849) [reply platform/containers/kubernetes-distro/install.sh]: but we don't have docker sock inside the container so shy keeping it?

**Pull request merged**